Add data import tasks

carbonin opened the pull request:
These tasks will run a job for each PV to move data from an existing location to the PV using rsync.

These tasks are included into the existing foreman.yml playbook for the start tag if the data_import variable is set to `true`.

An additional playbook is also provided to run the data import jobs outside of the foreman application deployment process.

To run a standalone import:
```
ansible-playbook -e import_source_host=foreman.example.com -e import_source_user=foreman import.yml
```

To run an import and start in the same command:
```
ansible-playbook -e data_import=true -e import_source_host=foreman.example.com -e import_source_user=foreman -t start foreman.yml
```

--------------------------------------------------------------------------------

The import container will use the provided user to rsync the data to a mounted PV. The application will later use the same PVC to access the previously synced data.
The deployment assumes that there is a secret in the project called `data-import-secret` which contains an ssh private key. The corresponding public key must have already been installed into the host specified in `import_source_host`

An example secret would be:
```yaml
apiVersion: v1
kind: Secret
metadata:
  name: data-import-secret
stringData:
  id_rsa: |-
    <key text goes here>
```

Leaving this as WIP for now until we can figure out how we want to deal with the ssh keys (e.g. do we need an example file? Should we generate the key pair in ansible?)

carbonin committed:
Add data import tasks

These tasks will run a job for each PV to move data from an existing
location to the PV using rsync.

These tasks are included into the existing foreman.yml playbook
for the start tag if the data_import variable is set to `true`.

An additional playbook is also provided to run the data import
jobs outside of the foreman application deployment process.

carbonin commented:
Also, requires https://github.com/ehelms/forklift/pull/17

carbonin commented on containers/deploy/import/jobs.yml:
```diff
@@ -0,0 +1,72 @@
+  - postgres
+  until: job["job"]["status"]["succeeded"] == 1
+  delay: 10
+  retries: 200
```
Really want this to be retry forever, but not sure how to do that.

ehelms replied:
Not sure other than a really, really big number.

carbonin commented on containers/deploy/import/jobs.yml:
```diff
@@ -0,0 +1,72 @@
+---
+- fail: msg="Import task requires \'import_source_host\' and \'import_source_user\' to be provided"
+  when: import_source_host is undefined or import_source_user is undefined
```
Are these names okay?

ehelms replied:
Seem fine.

carbonin commented on containers/deploy/import/jobs.yml:
```diff
@@ -0,0 +1,72 @@
+  - {name: 'pulp',     source: '/var/lib/pulp'}
+  - {name: 'mongodb',  source: '/var/lib/mongodb'}
+  - {name: 'puppet',   source: '/etc/puppetlabs/code/environments'}
+  - {name: 'postgres', source: '/var/lib/postgresql/data'}
```
@ehelms Can you double check these paths? Also, do these need to be configurable?

ehelms commented:
Code looks good.

carbonin commented:
Okay, so what do you think about the ssh key plan @ehelms ?
What do you think would be the most comfortable path for a user? 

I was thinking something like write some tasks to generate a keypair locally and then generate the secret with the private key, but then the user needs to install the public key onto their existing server so we wouldn't really be able to do it in one shot.

ehelms commented:
For the demo purpose, we could pre-generate the key just like we do the certificates. Optionally encrypt the private key using ansible-vault. In practice, I think we can provide facilities to generate the key but also allow a user, with instructions, how to populate the ansible secret/variable file with their own key to then get it into Openshift.

TLDR; lets go with ssh key creation role/playbook to generate them, but ensure a user could just supply their own by copying it into whatever Ansible variable file.

carbonin committed:
Add ssh keys and secret for data import

For testing a user can deploy the data import job using the private
key in secrets.yml and install the matching public key (import-key.pub.dev)
into the machine they want to import from.

carbonin committed:
Use correct path to puppet and postgres and use project name for jobs

carbonin committed:
Create the import secret from the foreman.yml playbook

carbonin committed:
Reformat to create the secret properly

Previously the secret would be created with a filename of
"idRsa" rather than "id_rsa"

carbonin commented:
Opened an issue about the behavior around how the secrets are created when the keys have underscore characters in them. https://github.com/ansible/ansible-kubernetes-modules/issues/33

We can leave this for now, or change the name of the private key file in the sync image. Thoughts?

carbonin committed:
Change wait time to 4 hours total for data import to complete

ehelms commented on containers/deploy/secrets/import_secrets.yml:
```diff
@@ -0,0 +1,14 @@
+              namespace: "{{ project_name }}"
+          type: Opaque
+      data:
+          id_rsa: '{{ import_key | b64encode }}'
```
I am missing where this ends up getting used via a volumeMount.

carbonin replied:
https://github.com/ehelms/forklift/pull/18/files#diff-795fb7a5c61e5ce8d7524ac21b2b2f7bR48
